numpy/core/src/npymath/ieee754.cpp

```diff
@@ -516,53 +549,100 @@ npy_longdouble npy_nextafterl(npy_longdouble x, npy_longdouble y)
 }
 #endif
 
-/**begin repeat
- * #suff = f,,l#
- * #SUFF = F,,L#
- * #type = npy_float, npy_double, npy_longdouble#
- */
-@type@ npy_spacing@suff@(@type@ x)
+namespace {
+template <typename T>
+struct numeric_limits;
+
+template <>
+struct numeric_limits<float> {
+    static const npy_float nan;
+};
+const npy_float numeric_limits<float>::nan = NPY_NANF;
+
+template <>
+struct numeric_limits<double> {
+    static const npy_double nan;
+};
+const npy_double numeric_limits<double>::nan = NPY_NAN;
+
+template <>
+struct numeric_limits<long double> {
+    static const npy_longdouble nan;
+};
+const npy_longdouble numeric_limits<long double>::nan = NPY_NANL;
+}  // namespace
+
+template <typename type>
+static type
+_npy_spacing(type x)
 {
     /* XXX: npy isnan/isinf may be optimized by bit twiddling */
     if (npy_isinf(x)) {
-        return NPY_NAN@SUFF@;
+        return numeric_limits<type>::nan;
     }
 
-    return _next@suff@(x, 1) - x;
+    return _next(x, 1) - x;
+}
+
+/*
+ * Instantiation of C interface
+ */
+extern "C" {
+npy_float
+npy_spacingf(npy_float x)
+{
+    return _npy_spacing(x);
+}
+npy_double
+npy_spacing(npy_double x)
+{
+    return _npy_spacing(x);
+}
+npy_longdouble
+npy_spacingl(npy_longdouble x)
+{
+    return _npy_spacing(x);
+}
```

Comment:
One of the CI tests runs into this:
```
numpy/core/src/npymath/ieee754.cpp:282:1: warning: ‘long double _next(long double, int)’ defined but not used [-Wunused-function]
  282 | _next(long double x, int p)
      | ^~~~~
```
I don't really understand why, since I feel `_npy_spacing` here should use `_next(long double)`?  My guess would be that this comes from the `long double` and `npy_longdouble` difference in the end.  I.e. I guess `npy_longdouble` may be defined as double on platforms where they are identical.  But C++ may still distnguish?
(The warning seems harmless in itself, but...)

Reply:
Ah, maybe it just doesn't matter much. I guess this CI probably just doesn't care about warnings and the failure is the other reason.